osu.Game/Database/RealmArchiveModelImporter.cs

```diff
@@ -105,22 +105,27 @@ public async Task<IEnumerable<Live<TModel>>> Import(ProgressNotification notific
             }
 
             notification.Progress = 0;
-            notification.Text = $"{HumanisedModelName.Humanize(LetterCasing.Title)} import is initialising...";
 
             int current = 0;
 
             var imported = new List<Live<TModel>>();
 
             parameters.Batch |= tasks.Length >= minimum_items_considered_batch_import;
 
-            await Task.WhenAll(tasks.Select(async task =>
+            notification.Text = $"{HumanisedModelName.Humanize(LetterCasing.Title)} import is initialising...";
+            notification.State = ProgressNotificationState.Active;
+
+            await pauseIfNecessaryAsync(parameters, notification, notification.CancellationToken).ConfigureAwait(false);
+
+            await Parallel.ForEachAsync(tasks, notification.CancellationToken, async (task, cancellation) =>
             {
-                if (notification.CancellationToken.IsCancellationRequested)
-                    return;
+                cancellation.ThrowIfCancellationRequested();
```

Comment:
Are you sure about this? Won't this be logged (to the user also) as an unobserved exception? I mean... see the try-catch below which even swallows the `OperationCanceledException`

Reply:
It definitely doesn't come up as an unobserved else I would have noticed.

You can test using:

```diff
diff --git a/osu.Game/Database/RealmArchiveModelImporter.cs b/osu.Game/Database/RealmArchiveModelImporter.cs
index c4eb93b754..addba3f338 100644
--- a/osu.Game/Database/RealmArchiveModelImporter.cs
+++ b/osu.Game/Database/RealmArchiveModelImporter.cs
@@ -119,6 +119,9 @@ public async Task<IEnumerable<Live<TModel>>> Import(ProgressNotification notific
 
             await Parallel.ForEachAsync(tasks, notification.CancellationToken, async (task, cancellation) =>
             {
+                while (!notification.CancellationToken.IsCancellationRequested)
+                    Thread.Sleep(1000);
+
                 cancellation.ThrowIfCancellationRequested();
 
                 try
@@ -139,6 +142,7 @@ await Parallel.ForEachAsync(tasks, notification.CancellationToken, async (task,
                 }
                 catch (OperationCanceledException)
                 {
+                    throw;
                 }
                 catch (Exception e)
                 {
@@ -531,7 +535,8 @@ protected virtual void PostImport(TModel model, Realm realm, ImportParameters pa
         /// <param name="model">The new model proposed for import.</param>
         /// <param name="realm">The current realm context.</param>
         /// <returns>An existing model which matches the criteria to skip importing, else null.</returns>
-        protected TModel? CheckForExisting(TModel model, Realm realm) => string.IsNullOrEmpty(model.Hash) ? null : realm.All<TModel>().OrderBy(b => b.DeletePending).FirstOrDefault(b => b.Hash == model.Hash);
+        protected TModel? CheckForExisting(TModel model, Realm realm) =>
+            string.IsNullOrEmpty(model.Hash) ? null : realm.All<TModel>().OrderBy(b => b.DeletePending).FirstOrDefault(b => b.Hash == model.Hash);
 
         /// <summary>
         /// Whether import can be skipped after finding an existing import early in the process.

```

Debugging through I see a `TaskCanceledException` but it gracefully handles somewhere (debugger doesn't let me see).

Reply:
It is actually causing an issue with your diff. When you cancel the notification the download/download buttons (in beatmap listing) remain active in an "importing" state, with no way to restore. 

https://github.com/ppy/osu/blob/5276f77ebd395b34fa5b7425067efde744567897/osu.Game/Database/ModelDownloader.cs#L67-L84

I'm seeing nothing special about what's happening here, it just looks like `OperationCanceledException`s are never unobserved / don't reach that handler. Throwing `InvalidOperationException` works, throwing `OperationCanceledException` doesn't 😅 

Also, it's actually even caused by _just_ passing into the `Parallel.ForEachAsync()` itself without throwing from inside the callback delegate...